serve/mlc_serve/model/tvm_model.py

```diff
@@ -0,0 +1,494 @@
+                                        sequence_id.request_id, seq_id
+                                    ),
+                                    generated_tokens=[],
+                                    error=err_msg,
+                                )
+                            )
+                    else:
+                        outputs.append(
+                            TextGenerationResult(
+                                sequence_id=sequence_id,
+                                generated_tokens=[],
+                                error=err_msg,
+                            )
+                        )
+
+            return outputs
+
+
+def init_tvm_model(
+    model_artifact_config: ModelArtifactConfig, engine_config: MLCServeEngineConfig
+) -> Tuple[TextGenerator, CacheManager]:
+    dev = tvm.device("cuda", 0)
+
+    model = Model(model_artifact_config, dev)
+
+    if model_artifact_config.num_shards > 1:
+        model.disco_session.sync_worker_0()
+
+    num_kv_heads = (
+        model_artifact_config.num_key_value_heads // model_artifact_config.num_shards
+    )
+    head_size = (
+        model_artifact_config.hidden_size // model_artifact_config.num_attention_heads
+    )
+
+    if engine_config.max_num_batched_tokens > 0:
+        LOG.info("Running memory profiling.")
+        num_blocks = get_num_cache_blocks(
+            model,
+            [engine_config.max_input_len] * engine_config.max_num_sequences,
+            model_artifact_config.num_hidden_layers,
+            num_kv_heads,
+            head_size,
+        )
+    else:
+        num_blocks = 500
+
+    num_cache_slots = num_blocks * CacheManager.block_size
+
+    if num_cache_slots <= engine_config.max_num_batched_tokens:
+        raise RuntimeError(
+            f"max_num_batched_tokens = {engine_config.max_num_batched_tokens} but"
+            f" only {num_blocks} cache blocks can be allocated. The number of"
+            f" available cache slots is {num_cache_slots}, not enough for"
+            f" {engine_config.max_num_batched_tokens} tokens. Try reducing"
+            " --max_input_len or --max_num_sequences."
+        )
+
+    LOG.info(f"Using {num_blocks} cache blocks.")
+
+    if model.disco_session:
+        init_cache_func = model.disco_session.get_global_func(
+            "tvm.contrib.vllm.allocate_kv_cache"
+        )
+    else:
+        init_cache_func = tvm.get_global_func("tvm.contrib.vllm.allocate_kv_cache")
+
+    cache_blocks = init_cache_func(
+        head_size,
+        model_artifact_config.num_hidden_layers,
+        num_kv_heads,
+        CacheManager.block_size,
+        num_blocks,
+    )
+
+    cache_manager = CacheManager(
+        cache_blocks,
+        num_blocks,
+        model_artifact_config.sliding_window,
+    )
+
+    LOG.info("Allocated KV cache blocks.")
+
+    # TODO(masahi): Make mypy understand that model confirms to TextGenerator Protocol.
+    return model, cache_manager  # type: ignore
```

Comment:
@yelite Any idea how to get protocol typing working here?

Reply:
There are two issues here:
1. `TextGenerator.generate` takes `Sequence[Union[PrefillRequest, DecodeRequest]]`, not `Union[List[PrefillRequest], List[DecodeRequest]]`.
2. The `kv_cache: KVCache` on `TextGenerator.generate` prevents the `model` here from conforming to the protocol. This could be solved by using a type parameter and make TextGenerator a generic protocol. But for the sake of simplicity, I think it's better to just use the Any type for `kv_cache` parameter.
```diff
diff --git a/serve/mlc_serve/engine/model_module.py b/serve/mlc_serve/engine/model_module.py
index 71d10c8e..3908db85 100644
--- a/serve/mlc_serve/engine/model_module.py
+++ b/serve/mlc_serve/engine/model_module.py
@@ -118,7 +118,7 @@ class TextGenerator(Protocol):
     def generate(
         self,
         requests: Sequence[Union[PrefillRequest, DecodeRequest]],
-        kv_cache: KVCache,
+        kv_cache,
     ) -> List[TextGenerationResult]:
         """
         A unified entrypoint for text generation.
diff --git a/serve/mlc_serve/model/tvm_model.py b/serve/mlc_serve/model/tvm_model.py
index 76e4451d..124d60d7 100644
--- a/serve/mlc_serve/model/tvm_model.py
+++ b/serve/mlc_serve/model/tvm_model.py
@@ -1,6 +1,6 @@
 import math
 import os
-from typing import List, Union, Tuple
+from typing import List, Union, Tuple, Sequence
 
 import structlog
 import numpy as np
@@ -203,7 +204,7 @@ class Model:
 
     def generate(
         self,
-        requests: Union[List[PrefillRequest], List[DecodeRequest]],
+        requests: Sequence[Union[PrefillRequest, DecodeRequest]],
         cache: KVCache,
     ) -> List[TextGenerationResult]:
         if len(requests) == 0:
@@ -491,4 +492,4 @@ def init_tvm_model(
     LOG.info("Allocated KV cache blocks.")
 
     # TODO(masahi): Make mypy understand that model confirms to TextGenerator Protocol.
-    return model, cache_manager  # type: ignore
+    return model, cache_manager 
```